packages/kbn-generate-csv/src/generate_csv_esql.ts

```diff
@@ -0,0 +1,221 @@
+ * Side Public License, v 1.
+ */
+
+import { lastValueFrom } from 'rxjs';
+import type { Writable } from 'stream';
+
+import { errors as esErrors } from '@elastic/elasticsearch';
+import type { IScopedClusterClient, IUiSettingsClient, Logger } from '@kbn/core/server';
+import type { ESQLSearchParams, ESQLSearchReponse } from '@kbn/es-types';
+import {
+  cellHasFormulas,
+  ESQL_SEARCH_STRATEGY,
+  getEsQueryConfig,
+  IKibanaSearchRequest,
+  IKibanaSearchResponse,
+} from '@kbn/data-plugin/common';
+import type { IScopedSearchClient } from '@kbn/data-plugin/server';
+import {
+  AuthenticationExpiredError,
+  byteSizeValueToNumber,
+  CancellationToken,
+  ReportingError,
+} from '@kbn/reporting-common';
+import type { TaskRunResult } from '@kbn/reporting-common/types';
+import type { ReportingConfigType } from '@kbn/reporting-server';
+import { buildEsQuery, Filter } from '@kbn/es-query';
+import { zipObject } from 'lodash';
+
+import { CONTENT_TYPE_CSV } from './constants';
+import { CsvExportSettings, getExportSettings } from './get_export_settings';
+import { i18nTexts } from './i18n_texts';
+import { MaxSizeStringBuilder } from './max_size_string_builder';
+
+export interface JobParamsCsvESQL {
+  query: { esql: string };
+  columns?: string[];
+  filters?: Filter[];
+  browserTimezone?: string;
+}
+
+interface Clients {
+  es: IScopedClusterClient;
+  data: IScopedSearchClient;
+  uiSettings: IUiSettingsClient;
+}
+
+export class CsvESQLGenerator {
+  private csvContainsFormulas = false;
+  private maxSizeReached = false;
+  private csvRowCount = 0;
+
+  constructor(
+    private job: JobParamsCsvESQL,
+    private config: ReportingConfigType['csv'],
+    private clients: Clients,
+    private cancellationToken: CancellationToken,
+    private logger: Logger,
+    private stream: Writable
+  ) {}
+
+  public async generateData(): Promise<TaskRunResult> {
+    const settings = await getExportSettings(
+      this.clients.uiSettings,
+      this.config,
+      this.job.browserTimezone,
+      this.logger
+    );
+
+    let reportingError: undefined | ReportingError;
+    const warnings: string[] = [];
+
+    const { maxSizeBytes, bom, escapeFormulaValues } = settings;
+    const builder = new MaxSizeStringBuilder(this.stream, byteSizeValueToNumber(maxSizeBytes), bom);
+
+    const filter =
+      this.job.filters &&
+      buildEsQuery(
+        undefined,
+        [],
+        this.job.filters,
+        getEsQueryConfig(this.clients.uiSettings as Parameters<typeof getEsQueryConfig>[0])
+      );
+
+    const searchParams: IKibanaSearchRequest<ESQLSearchParams> = {
+      params: {
+        query: this.job.query.esql,
+        filter,
+        // TODO: time_zone support was temporarily removed from ES|QL,
+        // we will need to add it back in once it is supported again.
+        // https://github.com/elastic/elasticsearch/pull/102767
+        // timezone
+        // TODO: pass locale from browser?
+        // locale
```

Comment:
This wasn't available in the locator, I wasn't eager to add it as wasn't sure if this was significant. Happy to add if needed.

Reply:
What would `locale` be used for? Is it different than timezone?

Reply:
https://github.com/elastic/kibana/pull/174511#issuecomment-1887037057

Reply:
Thanks! I wonder if we should not use field formatters for the non-ES|QL CSV export as well. Do you know if any plans are changing with field formatters?

Reply:
> Do you know if any plans are changing with field formatters?
don't know :(

> I wonder if we should not use field formatters for the non-ES|QL CSV export as well.
I thought that the idea was to try to match the Discover table as closely as possible. I think it makes sense to keep the formatters for non-ESQL until they are there in the Discover table.